### PR TITLE
feat(clusters): fetch clusters on demand for very large apps

### PR DESCRIPTION
**Pull request by anotherchrisberry**
Deck suffers significant performance issues when trying to load the server groups for very large applications - large in the sense of having many clusters, not many instances in several clusters.

The solution we've come up with is to fetch those clusters as requested. Backend bits are already in place.

The API call to get the list of clusters for an application is very fast, even with thousands of clusters, so we do that before trying to get the server groups. If we find more than 250 clusters, we switch to on-demand mode.

The clusters view looks like this:
screen shot 2017-04-08 at 4 15 29 pm

Clicking into the search box ("(cluster name here)" shows the actual cluster name IRL):
screen shot 2017-04-08 at 4 19 04 pm

With some selected clusters:
screen shot 2017-04-08 at 4 20 59 pm

Also converted the cluster service to TS and removed a bunch of cruft. I thought about doing that conversion in a separate PR, which would have been a good idea for the sake of reviewability, but it's too late now.

I considered converting other modules to TS (allClustersController, ClusterFilterModel, but...we're already creeping up on 2k LOC here. Some other day.

@jrsquared @christopherthielen PTAL

### CODE REVIEW

File: app/scripts/modules/core/cluster/task.matcher.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+import {has, intersection, map} from 'lodash';
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/cluster/task.matcher.ts]: I pulled these out of the original cluster service, but made it a plain old file instead of an Angular module, since there's nothing injected and I don't imagine we'll ever inject anything Angular into it, in light of recent events.

File: app/scripts/modules/core/cluster/cluster.service.ts [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+          clusters.push(cluster);
+        });
+      });
+    });
+    this.addProvidersAndServerGroupsToInstances(serverGroups);
+    return clusters;
+  }
+
+  public addExecutionsToServerGroups(application: Application): void {
+    let executions = get(application, 'runningExecutions.data', []);
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/cluster.service.ts]: const

File: app/scripts/modules/core/cluster/cluster.service.ts [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+  public addTasksToServerGroups(application: Application): void {
+    let runningTasks = get(application, 'runningTasks.data', []);
+    if (!application.serverGroups.data) {
+      return; // still run if there are no running tasks, since they may have all finished and we need to clear them.
+    }
+    application.serverGroups.data.forEach((serverGroup: ServerGroup) => {
+      if (!serverGroup.runningTasks) {
+        serverGroup.runningTasks = [];
+      }
+      serverGroup.runningTasks.length = 0;
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/cluster.service.ts]: A little pedantic, but this should be in an `else`.

File: app/scripts/modules/core/cluster/cluster.service.ts [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+      runningTasks.forEach(function(task) {
+        if (taskMatches(task, serverGroup)) {
+          serverGroup.runningTasks.push(task);
+        }
+      });
+    });
+  }
+
+  private getClusters(application: string): IPromise<IClusterSummary[]> {
+    return this.API.one('applications').one(application).one('clusters').get().then((clustersMap: {[account: string]: string[]}) => {
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/cluster.service.ts]: Should we start adding `.catch()` to these?

File: app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
+import {IComponentController, IComponentOptions, IScope, module} from 'angular';
+import {Application} from '../../application/application.model';
+import {IClusterSummary} from '../../domain/ICluster';
+
+require ('./onDemandClusterPicker.component.less');
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts]: import

File: app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
+import {IComponentController, IComponentOptions, IScope, module} from 'angular';
+import {Application} from '../../application/application.model';
+import {IClusterSummary} from '../../domain/ICluster';
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts]: use `core/...` imports

File: app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+require ('./onDemandClusterPicker.component.less');
+
+class OnDemandClusterPickerController implements IComponentController {
+  public application: Application;
+  public availableClusters: IClusterSummary[];
+  public lastSelection: string;
+  public totalClusterCount: number;
+  public optionTemplate: string = require('./onDemandOptionTemplate.html');
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts]: Since you already inlined the bigger template for the component, let's inline this one too.

File: app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+           typeahead-focus-on-select="false"
+           typeahead-on-select="$ctrl.selectCluster($item)"
+           typeahead-template-url="{{$ctrl.optionTemplate}}"
+           />
+  `;
+}
+
+export const ON_DEMAND_CLUSTER_PICKER_COMPONENT = 'spinnaker.core.cluster.onDemandClusterPicker.component';
+module(ON_DEMAND_CLUSTER_PICKER_COMPONENT, [
+  require('./../filter/clusterFilter.model.js'),
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts]: `core/...`

File: app/scripts/modules/core/cluster/task.matcher.ts (hunk 1 above)
Comment (jrsquared) [inline app/scripts/modules/core/cluster/task.matcher.ts]: 😍

**Comment by tomaslin** [comment]
when you select a cluster name, is it a precise match or a select2 type match? worried that having to drill through 1000 cluster names is not fun

File: app/scripts/modules/core/cluster/task.matcher.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+import {ITask} from '../task/task.read.service';
+import {ServerGroup} from '../domain/serverGroup';
+
+export interface ITaskMatcher {
+ (task: ITask, serverGroup: ServerGroup): boolean;
+}
+
+function createcopylastasgMatcher(task: ITask, serverGroup: ServerGroup): boolean {
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/task.matcher.ts]: I know why this isn't camelCase, but it still looks awful.

File: app/scripts/modules/core/cluster/task.matcher.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  const source = task.getValueFor('source'),
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/task.matcher.ts]: no type needed for this `.getValueFor()` but was needed for the rest?

File: app/scripts/modules/core/cluster/task.matcher.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        deployedServerGroups: {[region: string]: string[]} = task.getValueFor('deploy.server.groups'),
+        targetServerGroup: string = targetRegion && deployedServerGroups && deployedServerGroups[targetRegion] ? deployedServerGroups[targetRegion][0] : null,
+        sourceServerGroup: string = source.asgName,
+        sourceAccount: string = source.account,
+        sourceRegion: string = source.region;
+
+  if (serverGroup.account === targetAccount && serverGroup.region === targetRegion && serverGroup.name === targetServerGroup) {
+    return true;
+  }
+  return serverGroup.account === sourceAccount && serverGroup.region === sourceRegion && serverGroup.name === sourceServerGroup;
```
Comment (jrsquared) [inline app/scripts/modules/core/cluster/task.matcher.ts]: Pretty hard to follow. Combine these two returns, confusing having them separate. Or better yet, store results in descriptive consts and return the `||` of those. Something like:
```
const serverGroupInTarget = serverGroup.account === targetAccount && serverGroup.region === targetRegion && serverGroup.name === targetServerGroup;
const serverGroupInSource = serverGroup.account === sourceAccount && serverGroup.region === sourceRegion && serverGroup.name === sourceServerGroup;
return serverGroupInTarget || serverGroupInSource;
```

**Comment by jrsquared** [comment]
Looks great! Nothing glaringly obvious, just some style/readability concerns. Oh, and that Travis yells at you.

File: app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts (hunk 7 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts]: Unfortunately, this one can't be inlined. The directive (uib-typeahead) needs the key in the template cache - it doesn't accept a full template.

File: app/scripts/modules/core/cluster/task.matcher.ts (hunk 10 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/cluster/task.matcher.ts]: The compiler/linter doesn't complain because `.getValueFor()` returns `any`. I'll add an `any` here for consistency.

File: app/scripts/modules/core/cluster/task.matcher.ts (hunk 11 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/cluster/task.matcher.ts]: Sounds good

File: app/scripts/modules/core/cluster/task.matcher.ts (hunk 9 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/cluster/task.matcher.ts]: yep, agreed

**Comment by anotherchrisberry** [comment]
@tomaslin it's a select2-style - it filters/matches on whatever is typed in

File: app/scripts/modules/core/cluster/cluster.service.ts (hunk 4 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/cluster/cluster.service.ts]: Good catch - I need to work through this. Right now, if this call fails, the page won't load at all, even to indicate that there was an error. I'll get this fixed and updated later this evening.

File: app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts (hunk 8 above)
Comment (christopherthielen) [inline app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts]: What is the rule for importing from `core` vs `../`?

File: app/scripts/modules/core/cluster/cluster.service.ts [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+            stageRegion === serverGroup.region) {
+              serverGroup.runningExecutions.push(execution);
+          }
+        });
+      });
+    });
+  }
+
+  public addTasksToServerGroups(application: Application): void {
+    let runningTasks = get(application, 'runningTasks.data', []);
```
Comment (christopherthielen) [inline app/scripts/modules/core/cluster/cluster.service.ts]: Should we just add the linter rule for this?

https://palantir.github.io/tslint/rules/prefer-const/

File: app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts (hunk 8 above)
Comment (jrsquared) [inline app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts]: I don't think there's a rule, will defer to Chris, but I don't like seeing ../ anywhere. A class shouldn't rely on its parents path structure.

File: app/scripts/modules/core/cluster/cluster.service.ts (hunk 12 above)
Comment (jrsquared) [inline app/scripts/modules/core/cluster/cluster.service.ts]: YES. PLEASE.

**Comment by tomaslin** [comment]
looks great

File: app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+           typeahead-on-select="$ctrl.selectCluster($item)"
+           typeahead-template-url="{{$ctrl.optionTemplate}}"
+           />
+  `;
+}
+
+export const ON_DEMAND_CLUSTER_PICKER_COMPONENT = 'spinnaker.core.cluster.onDemandClusterPicker.component';
+module(ON_DEMAND_CLUSTER_PICKER_COMPONENT, [
+  require('../filter/clusterFilter.model.js'),
+]).component('onDemandClusterPicker', new OnDemandClusterPickerComponent());
```
Comment (christopherthielen) [inline app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts]: Does using a `class` buy us anything?  If not, my preference is to use POJOs for component defs:

```
const template = `
    <input type="text"
           class="form-control"
           placeholder="{{$ctrl.totalClusterCount}} clusters found in this application. Select a cluster to view"
           ng-model="$ctrl.lastSelection"
           uib-typeahead="cluster for cluster in $ctrl.availableClusters | filter: $viewValue | limitTo: 50"
           typeahead-min-length="0"
           typeahead-focus-on-select="false"
           typeahead-on-select="$ctrl.selectCluster($item)"
           typeahead-template-url="{{$ctrl.optionTemplate}}"
           />
`;

const onDemandClusterPickerComponent: IComponentOptions = {
  template: template,
  controller: OnDemandClusterPickerController,
  bindings: {
    application: '<',
  },
}

export const ON_DEMAND_CLUSTER_PICKER_COMPONENT = 'spinnaker.core.cluster.onDemandClusterPicker.component';
module(ON_DEMAND_CLUSTER_PICKER_COMPONENT, [
  require('../filter/clusterFilter.model.js'),
]).component('onDemandClusterPicker', onDemandClusterPickerComponent());
```

File: app/scripts/modules/core/cluster/task.matcher.ts (hunk 9 above)
Comment (christopherthielen) [inline app/scripts/modules/core/cluster/task.matcher.ts]: I don't know why this isn't camelCase.  Care to explain?

Comment (anotherchrisberry) [inline app/scripts/modules/core/cluster/task.matcher.ts]: The matcher names are keyed off the `notification.type` field in the task context, which is all lower case. No idea why we decided to make them all lower case, but we did. Newer ones, e.g. `rollbackServerGroup`, is camel cased. Go figure.

The `function`s in this file could be camel cased, but I decided to leave them in the same case as the value they're matching, just to be sort of consistent.

File: app/scripts/modules/core/cluster/cluster.service.ts (hunk 4 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/cluster/cluster.service.ts]: After playing around with this, I don't think we want a `.catch()` here. If the call fails (e.g. gate returns a non-200 status), the `getList()` call will reject. I tested through the call stack of the calling method, throwing random exceptions in every possible place, and it ended up with us in a rejected state, which is caught by the `applicationDataSource` and handled appropriately: it sets the `loadFailure` flag on the data source and rejects the `.ready()` promise (or the `.refresh()` promise, depending on which is called).

Basically, I think we're good here. Adding a `.catch` would just obfuscate the error - at least in the case of application data source loading.

**Comment by jrsquared** [comment]
Perfect. As long as we're failing appropriately, unlike when server groups
429s. :flushed:

File: app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts (hunk 13 above)
Comment (anotherchrisberry) [inline app/scripts/modules/core/cluster/onDemand/onDemandClusterPicker.component.ts]: I'm not opposed to that - will try it out and see how it goes (we can always switch back to a class if there's a reason to do so that I don't remember).
